### PR TITLE
Add back --dist-dir and --public-url CLI options

### PR DESCRIPTION
**Pull request by devongovett**
Previously, we had removed the `--out-dir` and `--public-url` CLI options that existed in Parcel 1 due to multi-target support in Parcel 2. However, due to confusion, along with needs to change these dynamically rather than hard coding them in package.json, they make their triumphant return in this PR.

`publicUrl` and `distDir` are still options to targets, but the CLI options are now used by default in case these are not defined. In particular, `--dist-dir` is used when there isn't a top-level package.json field for the target's main entry, and `--public-url` is used when a `publicUrl` option isn't set in the target config in package.json.

### CODE REVIEW

File: packages/core/core/src/AssetGraphBuilder.js [hunk 1]
```diff
@@ -79,11 +79,13 @@ export default class AssetGraphBuilder extends EventEmitter {
     this.options = options;
     this.assetRequests = [];
 
-    let {hot} = options;
+    // TODO: changing these should not throw away the entire graph.
+    // We just need to re-run target resolution.
+    let {hot, publicUrl, distDir, minify, scopeHoist} = options;
```
Comment (devongovett) [inline packages/core/core/src/AssetGraphBuilder.js]: Should we have a node in the request graph to invalidate based on options? @padmaia

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 4.01s | -1.11s 🚀 |
| Cached | 3.60s | -46.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 1.07kb | +5.00b ⚠️ | 231.00ms | -0.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/parcel.a2ee7a0b.webp | 102.94kb | +0.00b | 13.00ms | -374.00ms 🚀 |
| dist/index.js | 1.07kb | +5.00b ⚠️ | 17.00ms | -375.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.25s | -125.00ms |
| Cached | 3.13s | -450.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 64.04kb | +397.00b ⚠️ | 1.20s | -1.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 64.04kb | +397.00b ⚠️ | 15.00ms | -381.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/ak-editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 4.27s | +80.00ms |
| Cached | 3.48s | -187.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 743.00b | +7.00b ⚠️ | 192.00ms | +3.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 743.00b | +7.00b ⚠️ | 14.00ms | +1.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/257130351657222666)**

File: packages/core/core/src/AssetGraphBuilder.js (hunk 1 above)
Comment (padmaia) [reply packages/core/core/src/AssetGraphBuilder.js]: Yeah, that sounds like a good idea.

**Comment by fregante** [comment]
Why not just keep it as `--out-dir`? Related: https://github.com/parcel-bundler/parcel/issues/3993

**Pull request merged**